src/docmanager/default.ts

```diff
@@ -0,0 +1,175 @@
+// Copyright (c) Jupyter Development Team.
+// Distributed under the terms of the Modified BSD License.
+'use strict';
+
+import {
+  IKernelSpecId
+} from 'jupyter-js-services';
+
+import {
+  ISignal, Signal
+} from 'phosphor-signaling';
+
+import {
+  loadModeByFileName
+} from '../codemirror';
+
+import {
+  CodeMirrorWidget
+} from '../codemirror/widget';
+
+import {
+  IDocumentModel, IWidgetFactory, IModelFactory, IDocumentContext
+} from './index';
+
+
+/**
+ * The default implementation a document model.
+ */
+export
+class DocumentModel implements IDocumentModel {
+  /**
+   * Construct a new document model.
+   */
+  constructor(path: string, kernelSpecs: IKernelSpecId[]) {
+    // Use the path and the kernel specs to set the default
+    // kernel and language.
+  }
```

Comment:
Why do you need to know what kernels are available? I saw the preferred kernel as an intrinsic property of the file/filename.